### ISSUE TITLE
New rule to warn about the usage of this keyword inside static class methods

### ISSUE THREAD

**Issue by bertolo1988** [eslint/eslint#11234]
Take a look at this silly class

```js
class Rectangle {
  constructor(width, height) {
    this.width = width;
    this.height = height;
  }
  getArea() {
    return this.width * this.height;
  }
  static getHeight() {
    return this.height();
  }
}
```

This is valid JS while it has a lot of danger in it. I suggest the creation of a rule that prevents people from calling `this` from inside static class methods.

I would even suggest that eslint default config should give a warning on this.

**Comment by ljharb** [comment]
While it calling an instance method on the constructor is obviously silly, it’s not a problem with `this`. Using `this` in static methods is completely normal and correct when needed, to allow subclasses to override static methods on the constructor as well.

**Comment by bertolo1988** [comment]
> Using this in static methods is completely normal and correct when needed, to allow subclasses to override static methods on the constructor as well.

Can you give an example?

**Comment by ljharb** [comment]
```js
class Parent {
  static a(){
    return this.b();
  }
  static b() {
    return false;
  }
}

class Child extends Parent {
  static b() {
    return true;
  }
}

Parent.a() // false
Child.a() // true
```

**Comment by bertolo1988** [comment]
> however, that isn’t directly related to the use of “this” in a static method

Imo a static method, in any language, should be stateless and only allowed to access constants, other static methods, and provided arguments. Well, I this is another discussion... JS is not Java indeed.



>  Would a rule that warned on static/instance mismatches cover your use cases?

Yes, agreed. This should be the main objective of that said new rule.

> What would that rule look like?

My suggestion would be a rule that warns developers of:

* the usage of instance methods inside static methods called upon the `this` keyword

* the usage of static methods inside instance methods called upon the `this` keyword

What do you think?

**Comment by bertolo1988** [comment]
> but how would the rule behave with computed method names

Such as `Parent['myMethod']`? Are there any other ways?
I think this case should be taken into consideration too. Aren't there any other rules doing something similar?

>  including Symbols

I think Symbols have what it takes to be seen as constants. I'd have to do some more research on this.

> What about getters and setters

Should be seen as instance methods.

**Comment by ljharb** [comment]
note that you can have static getters and setters too.

By computed, i mean like:
```js
[anyExpressionHereIncludingVariables]() {}
```

**Comment by bertolo1988** [comment]
> note that you can have static getters and setters too.

True. That one is easier then. 😅

> [anyExpressionHereIncludingVariables]() {}

I would recommend to follow the same approach eslint uses in similar situations since I don't really know if there are similar rules.

I would be ok if the linter didn't say anything in this case.

**Comment by not-an-aardvark** [comment]
I think it could be useful to have a rule that warns when:

* A class does not have a superclass (i.e. it doesn't use the `extends` keyword)
* `this.someMethod()` is used in a static method of that class
* `someMethod` is not a static method of the class, and there are no static computed method names

That seems like it would be able to catch most error cases, without any false positives except when methods are added to a class after-the-fact.

**Comment by hornta** [comment]
> note that you can have static getters and setters too.
> 
> By computed, i mean like:
> 
> ```js
> [anyExpressionHereIncludingVariables]() {}
> ```

Do you know if there rules that disallow having computed properties like this?

**Comment by bertolo1988** [comment]
@hornta seems to me that such rule would be very "disruptive"

**Comment by ljharb** [comment]
Since that syntax is required for symbols, a first-class language feature, i don’t think it makes sense to ever ban computed properties.
